File: app/src/main/java/org/wikipedia/page/GeoUnmarshaller.java
package org.wikipedia.page;

import android.location.Location;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONException;
import org.json.JSONObject;

public final class GeoUnmarshaller {
    static final String LATITUDE = "lat";
    static final String LONGITUDE = "lon";

    @Nullable
    public static Location unmarshal(@Nullable String json) {
        if (json == null) {
            return null;
        }

        JSONObject jsonObj;
        try {
            jsonObj = new JSONObject(json);
        } catch (JSONException e) {
            return null;
        }
        return unmarshal(jsonObj);
    }

    @NonNull
    public static Location unmarshal(@NonNull JSONObject jsonObj) {
        Location ret = new Location((String) null);
        ret.setLatitude(jsonObj.optDouble(LATITUDE));
        ret.setLongitude(jsonObj.optDouble(LONGITUDE));
        return ret;
    }

    private GeoUnmarshaller() { }
}

File: app/src/main/java/org/wikipedia/page/GeoUnmarshaller.kt
package org.wikipedia.page

import android.location.Location
import org.json.JSONException
import org.json.JSONObject

object GeoUnmarshaller {
    const val LATITUDE = "lat"
    const val LONGITUDE = "lon"

    @JvmStatic
    fun unmarshal(json: String?): Location? {
        return json?.let {
            val jsonObj = try {
                JSONObject(it)
            } catch (e: JSONException) {
                return null
            }
            unmarshal(jsonObj)
        }
    }

    fun unmarshal(jsonObj: JSONObject): Location {
        return Location("").apply {
            latitude = jsonObj.optDouble(LATITUDE)
            longitude = jsonObj.optDouble(LONGITUDE)
        }
    }
}
